disktable.go

```diff
@@ -0,0 +1,650 @@
+	if _, err := os.Stat(pathDir); err != nil {
+		return nil, fmt.Errorf("pathDir %q does not exist", pathDir)
+	}
+
+	table := &Table{indexByName: map[string]*Index{}, logger: nullLogger{}}
+
+	if err := calloptions.ApplyOptions(table, options); err != nil {
+		return nil, err
+	}
+
+	td := tableDef{}
+	if err := td.unmarshal(pathDir); err != nil {
+		return nil, err
+	}
+	counter := atomic.Uint64{}
+	counter.Store(td.Length)
+
+	// Open all our index tables.
+	for _, col := range td.Columns {
+		badgeOpts := readOpts.WithDir(col.Path).WithValueDir(col.Path).WithLogger(table.logger)
+		db, err := badger.Open(badgeOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		index := &Index{
+			Name:            col.Name,
+			AllowDuplicates: col.AllowDuplicates,
+			db:              db,
+			buffPool:        make(chan *z.Buffer, 100),
+			counter:         counter,
+		}
+		table.indexes = append(table.indexes, index)
+		table.indexByName[col.Name] = index
+	}
+
+	// Open our primary data table.
+	primaryPath := filepath.Join(pathDir, "primary")
+	badgeOpts := readOpts.WithDir(primaryPath).WithValueDir(primaryPath).WithLogger(table.logger)
+	db, err := badger.Open(badgeOpts)
+	if err != nil {
+		return nil, err
+	}
+	table.primary = &Index{
+		Name:     "primary",
+		db:       db,
+		buffPool: make(chan *z.Buffer, 100),
+		counter:  counter,
+	}
+
+	return table, nil
+}
+
+// Get gets the i'th entry stored in the table.
+func (t *Table) Get(ctx context.Context, i uint64) ([]byte, error) {
+	// The primary index is stored with the starting index as 1. Externally
+	// we use the standard starting index of 0. So we i++.
+	i++
+
+	var v []byte
+	err := t.primary.db.View(
+		func(txn *badger.Txn) error {
+			item, err := txn.Get(NumToByte(i))
+			if err != nil {
+				return err
+			}
+
+			v, err = item.ValueCopy(nil)
+			if err != nil {
+				return fmt.Errorf("problem getting value at index %d: %w", i, err)
+			}
+			return nil
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+// Len() returns the number of entries in the table.
+func (t *Table) Len() uint64 {
+	return t.primary.counter.Load()
+}
+
+// Lookup provides the Index name and the Value that needs to match for the entry
+// to be returned.
+type Lookup struct {
+	// IndexName is the name of the index to do the lookup in.
+	IndexName string
+	// Key is the key in the index to lookup.
+	Key []byte
+}
+
+// Result is the result of a table lookup.
+type Result struct {
+	Value []byte
+	Err   error
+}
+
+type btreeItem struct {
+	primaryKey uint64
+	count      uint16
+}
+
+func (a btreeItem) Less(than btree.Item) bool {
+	b := than.(btreeItem)
+	return a.primaryKey < b.primaryKey
+}
+
+// Fetch retrieves specifc rows that match all index lookups. You must supply at least
+// 1 lookup. You cannot currently specify multiple searches in the same index.
+// If you wish to fetch all rows, use FetchAll().
+func (t *Table) Fetch(ctx context.Context, lookup ...Lookup) (chan Result, error) {
```

Comment:
since there needs to be >=1 `Lookup`, WDYT about this:

```suggestion
func (t *Table) Fetch(ctx context.Context, lookup1 Lookup, lookups ...Lookup) (chan Result, error) {
```

Reply:
Made the change, just slightly different names.